api/workload/x509_client.go

```diff
@@ -42,40 +56,164 @@ func NewX509Client(c *X509ClientConfig) *x509Client {
 	}
 
 	return &x509Client{
-		c:        c,
-		stopChan: make(chan struct{}),
-		stream:   newX509Stream(c),
+		c:          c,
+		updateChan: make(chan *workload.X509SVIDResponse, 1),
 	}
 }
 
 type x509Client struct {
-	c        *X509ClientConfig
-	stopChan chan struct{}
+	c          *X509ClientConfig
+	updateChan chan *workload.X509SVIDResponse
+	cancel     func()
 
-	stream *x509Stream
+	mu      sync.RWMutex
+	current *workload.X509SVIDResponse
 }
 
 func (x *x509Client) Start() error {
-	errChan := make(chan error, 1)
-	go func() { errChan <- x.stream.listen() }()
-	defer x.stream.stop()
-
-	select {
-	case <-x.stopChan:
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// make sure the client hasn't already been started
+	x.mu.Lock()
+	if x.cancel != nil {
+		x.mu.Unlock()
+		return errors.New("already started")
+	}
+	x.cancel = cancel
+	x.mu.Unlock()
+
+	// allow for another start after this function returns
+	defer func() {
+		x.mu.Lock()
+		x.cancel = nil
+		x.mu.Unlock()
+	}()
+
+	// set up a channel to receive the updates, update the current
+	// update, and push it down the update channel. the channel will
+	// be closed when this function returns to terminate this goroutine.
+	out := make(chan *workload.X509SVIDResponse)
+	defer close(out)
+	go func() {
+		for upd := range out {
+			// set the current update
+			x.mu.Lock()
+			x.current = upd
+			x.mu.Unlock()
+
+			// throw away an existing unread update
+			select {
+			case <-x.updateChan:
+			default:
+			}
+
+			// push the update down the channel
+			select {
+			case x.updateChan <- upd:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	err := streamX509SVID(ctx, x.c, out)
+	switch err {
+	case nil, context.Canceled:
 		return nil
-	case err := <-errChan:
+	default:
 		return err
 	}
 }
 
 func (x *x509Client) Stop() {
-	close(x.stopChan)
+	x.mu.RLock()
+	if x.cancel != nil {
+		x.cancel()
+	}
+	x.mu.RUnlock()
 }
 
 func (x *x509Client) CurrentSVID() (*workload.X509SVIDResponse, error) {
-	return x.stream.current()
+	x.mu.RLock()
+	defer x.mu.RUnlock()
+	if x.current == nil {
+		return nil, errors.New("no SVID received yet")
+	}
+	return x.current, nil
 }
 
 func (x *x509Client) UpdateChan() <-chan *workload.X509SVIDResponse {
-	return x.stream.updateChan()
+	return x.updateChan
+}
+
+func streamX509SVID(ctx context.Context, config *X509ClientConfig, out chan<- *workload.X509SVIDResponse) error {
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Pairs("workload.spiffe.io", "true"))
+
+	conn, err := dial.Dial(ctx, config.Addr)
+	if err != nil {
+		// dial shouldn't fail unless there is a misconfiguration
```

Comment:
I think dial should be backed off on also? Socket can be missing if agent hasn't yet started, or is in the middle of a restart

Reply:
grpc.Dial() is by default non-blocking and won't return connection attempt failures here unless you make it blocking (via grpc.WithBlock()).

Relevant documentation: https://godoc.org/google.golang.org/grpc#DialContext

From what I can tell, it will return a failure if the address you pass it is bogus, and in other circumstances (that we don't enable).

Reply:
> it will return a failure if the address you pass it is bogus

Is non-existent UDS considered bogus? CLI tool fails immediately on that

Reply:
You will get an immediate failure, just not from Dial(). It will come later when you make the request.

Reply:
(unless you pass `grpc.FailFast(false)` (deprecated) or `grpc.WaitUntilReady()`, and then it will happily retry with backoff).

Reply:
👍 thanks for the clarification